File: 13.gRPC-and-Inter-Service-Communication-for-Microservices/code/typescript/interceptors_the_middleware_of_grpc_8.ts
import * as grpc from '@grpc/grpc-js';

// Node.js gRPC handles interceptors quite differently from Go. 
// However, in @grpc/grpc-js there is interceptor support for clients, 
// and middleware-like patterns for servers (often through wrappers).
// We'll show a conceptual interceptor wrapper for a server method, 
// or how client interceptors work.

// A unary server interceptor concept:
function authInterceptor(
    call: grpc.ServerUnaryCall<any, any>,
    callback: grpc.sendUnaryData<any>,
    next: Function
) {
    const md = call.metadata;
    const tokens = md.get("authorization");
    
    if (tokens.length === 0 || !valid(tokens[0])) {
        return callback({
            code: grpc.status.UNAUTHENTICATED,
            message: "missing or invalid token",
            name: "Error"
        }, null);
    }
    
    // attach the verified identity for the handler to read from call
    (call as any).user = parse(tokens[0]);
    
    // call the next link / the real handler
    next(call, callback);
}

function loggingInterceptor(
    call: grpc.ServerUnaryCall<any, any>,
    callback: grpc.sendUnaryData<any>,
    next: Function
) {
    const start = Date.now();
    
    // intercept the callback
    const wrappedCallback = (err: any, value: any) => {
        const duration = Date.now() - start;
        console.log(`Call took ${duration}ms -> code: ${err ? err.code : 0}`);
        callback(err, value);
    };
    
    next(call, wrappedCallback);
}

// In practice, you wrap your handlers when building the server:
// Register the chain when building the server (outermost listed first):
function applyInterceptors(handler: Function, ...interceptors: Function[]) {
    return interceptors.reduceRight(
        (next, interceptor) => (call: any, cb: any) => interceptor(call, cb, next),
        handler
    );
}

// Example of binding it
// server.addService(UserServiceService, {
//    getUser: applyInterceptors(realGetUser, loggingInterceptor, authInterceptor)
// });

// Dummies
function valid(t: any): boolean { return true; }
function parse(t: any): any { return {}; }

File: 13.gRPC-and-Inter-Service-Communication-for-Microservices/code/rust/interceptors_the_middleware_of_grpc_8.rs
use tonic::{Request, Status, service::Interceptor};

// tonic interceptors run before the request is routed to the handler

// A unary server interceptor: signature is fixed by the framework.
fn auth_interceptor(mut req: Request<()>) -> Result<Request<()>, Status> {
    match req.metadata().get("authorization") {
        Some(t) if valid(t) => {
            // attach the verified identity for the handler to read from ctx
            let user = parse(t);
            req.extensions_mut().insert(user);
            Ok(req) // call the next link / the real handler
        }
        _ => Err(Status::unauthenticated("missing or invalid token")),
    }
}

// For logging in Tonic, you usually use a `tower` middleware layer instead 
// of an interceptor, because interceptors can't easily run code *after* the handler.
// Here's the tower middleware concept:
//
// let layer = tower::ServiceBuilder::new()
//     .layer(tower_http::trace::TraceLayer::new_for_grpc())
//     .into_inner();

// Register the chain when building the server (outermost listed first):
// let svc = UserServiceServer::with_interceptor(MyServer::default(), auth_interceptor);
// Server::builder()
//     .layer(layer)
//     .add_service(svc)
//     .serve(addr)
//     .await?;

// Dummies
fn valid(t: &tonic::metadata::MetadataValue<tonic::metadata::Ascii>) -> bool { true }
#[derive(Clone, Copy, Send, Sync, 'static)]
pub struct UserKey;
fn parse(t: &tonic::metadata::MetadataValue<tonic::metadata::Ascii>) -> UserKey { UserKey }
